cutadapt/writers.py

```diff
@@ -0,0 +1,185 @@
+"""
+Classes for managing output files.
+"""
+import os
+from . import seqio
+
+import sys
+UNCLOSEABLE = (sys.stdout, sys.stderr)
+
+
+class SingleEndWriter(object):
+	"""
+	Wrapper around a single output file. Can safely wrap
+	sys.stdout and sys.stderr.
+	"""
+	def __init__(self, handle):
+		self.handle = handle
+		self.written = 0
+		self.read1_bp = 0
+		self.read2_bp = 0
+	
+	def write(self, read1, read2=None):
+		assert read2 is None
+		self.handle.write(read1)
+		self.written += 1
+		self.read1_bp += len(read1)
+	
+	@property
+	def written_bp(self):
+		"""
+		Returns the total number of read1 and read2 bases
+		written to the output file.
+		"""
+		return (self.read1_bp, self.read2_bp)
+	
+	def close(self):
+		if self.handle not in UNCLOSEABLE:
+			self.handle.close()
+
+
+class PairedEndWriter(SingleEndWriter):
+	"""
+	Wrapper around a pair of output files.
+	"""
+	def write(self, read1, read2):
+		self.handle.write(read1, read2)
+		self.written += 1
+		self.read1_bp += len(read1)
+		self.read2_bp += len(read2)
+
+
+class Writers(object):
+	"""
+	Manage output files. Each output file (single-end) or pair of files
+	(paired-end) is associated with a filter type. Multiplexing is also
+	supported, in which case a filter type of `None` indicates that the
+	name of the output file to be used is determined by replacing '{name}'
+	in `name_pattern` with the adapter name. Files are opened lazily so 
+	that a `Writers` object can be constructed in the main thread and 
+	then be safely passed to a separate writer thread.
+	"""
+	def __init__(self, multiplexed=False, name_pattern=None, **seqio_open_args):
+		"""
+		Create a Writers object.
+		
+		multiplexed -- whether the default writer should be muliplexed,
+		  i.e. a separate file will be opened for each adapter name. If True, 
+		  `name_pattern` must contain '{name}', which will be replaced with the
+		  adapter name.
+		name_pattern -- The file name pattern for multiplexed files; ignored 
+		  unless `multiplexed` = True
```

Comment:
Good comments Marcel, Thanks. I’ve submitted fixes for all of these.

On Apr 13, 2016, at 9:53 AM, Marcel Martin <notifications@github.com> wrote:

> In cutadapt/writers.py:
> 
> > +	supported, in which case a filter type of `None` indicates that the
> > +	name of the output file to be used is determined by replacing '{name}'
> > +	in `name_pattern` with the adapter name. Files are opened lazily so 
> > +	that a `Writers` object can be constructed in the main thread and 
> > +	then be safely passed to a separate writer thread.
> > +	"""
> > +	def __init__(self, multiplexed=False, name_pattern=None, **seqio_open_args):
> > +		"""
> > +		Create a Writers object.
> > +		
> > +		multiplexed -- whether the default writer should be muliplexed,
> > +		  i.e. a separate file will be opened for each adapter name. If True, 
> > +		  `name_pattern` must contain '{name}', which will be replaced with the
> > +		  adapter name.
> > +		name_pattern -- The file name pattern for multiplexed files; ignored 
> > +		  unless `multiplexed` = True
> The multiplexed parameter isn’t needed: Just detect whether name_pattern is not None.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>